[lldb] update lldb-server platform help parsing

cs01 opened the pull request:
The lldb-server platform help text is inconsistent with lldb-server gdbserver help text. This PR modernizes the platform server to use LLVM's [TableGen](https://llvm.org/docs/TableGen/)-based option parsing (matching the existing gdbserver implementation), which auto-generates option parsing code and help text. 

The changes improve documentation quality by adding comprehensive option descriptions,, adding support for `-h`/`--help` flags, and organizing help output with DESCRIPTION and EXAMPLES sections. Internal-only options (`--child-platform-fd`) and unused legacy options (`--debug`, `--verbose`) are now hidden from help while maintaining backward compatibility. All functional behavior remains unchanged—this is purely a documentation and code modernization improvement.

## before
```
> /opt/llvm/bin/lldb-server p -h
p: unrecognized option '-h'
Usage:
  /opt/llvm/bin/lldb-server p [--log-file log-file-name] [--log-channels log-channel-list] [--port-file port-file-path] --server --listen port
```

## after
```
lldb-server p -h 
OVERVIEW: lldb-server platform

USAGE: lldb-server p [options] --listen <[host]:port> [[--] program args...]

CONNECTION OPTIONS:
  --gdbserver-port <port> Port to use for spawned gdbserver instances. If 0 or unspecified, a port will be chosen automatically. Short form: -P
  --listen <[host]:port>  Host and port to listen on. Format: [host]:port or protocol://[host]:port (e.g., tcp://localhost:1234, unix:///path/to/socket). Short form: -L
  --socket-file <path>    Write listening socket information (port number for TCP or path for Unix domain sockets) to the specified file. Short form: -f

GENERAL OPTIONS:
  --help            Display this help message and exit.
  --log-channels <channel1 categories...:channel2 categories...>
                    Channels to log. A colon-separated list of entries. Each entry starts with a channel followed by a space-separated list of categories. Common channels: lldb, gdb-remote, platform, process. Short form: -c
  --log-file <file> Destination file to log to. If empty, log to stderr. Short form: -l
  --server          Run in server mode, accepting multiple client connections sequentially. Without this flag, the server exits after handling the first connection.

OPTIONS:
  -- program args Arguments to pass to launched gdbserver instances.

DESCRIPTION
  Acts as a platform server for remote debugging. When LLDB clients connect,
  the platform server handles platform operations (file transfers, process
  launching) and spawns debug server instances (lldb-server gdbserver) to
  handle actual debugging sessions.

  By default, the server exits after handling one connection. Use --server
  to keep running and accept multiple connections sequentially.

EXAMPLES
  # Listen on port 1234, exit after first connection
  lldb-server platform --listen tcp://0.0.0.0:1234

  # Listen on port 5555, accept multiple connections
  lldb-server platform --server --listen tcp://localhost:5555

  # Listen on Unix domain socket
  lldb-server platform --listen unix:///tmp/lldb-server.sock

```

For comparison, here is the **gdbserver** help text:
```
lldb-server g -h
OVERVIEW: lldb-server

USAGE: lldb-server g[dbserver] [options] [[host]:port] [[--] program args...]

CONNECTION:
  --fd <fd>           Communicate over the given file descriptor.
  --named-pipe <name> Write port lldb-server will listen on to the given named pipe.
  --pipe <fd>         Write port lldb-server will listen on to the given file descriptor.
  --reverse-connect   Connect to the client instead of passively waiting for a connection. In this case [host]:port denotes the remote address to connect to.

GENERAL OPTIONS:
  --help            Prints out the usage information for lldb-server.
  --log-channels <channel1 categories...:channel2 categories...>
                    Channels to log. A colon-separated list of entries. Each entry starts with a channel followed by a space-separated list of categories.
  --log-file <file> Destination file to log to. If empty, log to stderr.
  --setsid          Run lldb-server in a new session.

TARGET SELECTION:
  --attach <pid-or-name> Attach to the process given by a (numeric) process id or a name.
  -- program args        Launch program for debugging.

DESCRIPTION
  lldb-server connects to the LLDB client, which drives the debugging session.
  If no connection options are given, the [host]:port argument must be present
  and will denote the address that lldb-server will listen on. [host] defaults
  to "localhost" if empty. Port can be zero, in which case the port number will
  be chosen dynamically and written to destinations given by --named-pipe and
  --pipe arguments.

  If no target is selected at startup, lldb-server can be directed by the LLDB
  client to launch or attach to a process.
```

cs01 committed:
update lldb-server platform help text

JDevlieghere reviewed (approved):
LGTM. I added some other active contributors to lldb-server for visibility.

DavidSpickett reviewed (commented):
I like the direction, and making `--help` actually work, surprising how many applications fall back on the "unknown option" message for it.

cs01 commented on lldb/test/Shell/lldb-server/TestPlatformHelp.test:
```diff
@@ -0,0 +1,41 @@
+RUN: %lldb-server platform -h 2>&1 | FileCheck %s
+
+CHECK: OVERVIEW: lldb-server platform
+CHECK: USAGE: lldb-server {{p|platform}} [options] --listen <{{[[]}}host]:port> {{[[]}}[--] program args...]
```
It was kinda hard to read, it was basically looking for a `[` but was put inside `[]`. I changed it to be `\[` so now it's `{{\[}}`. Hopefully it is a bit easier to read.

DavidSpickett replied:
I think you forgot to update the changes, but in any case:
* I did not realise that's how regex character `[]` worked, TIL.
* Couldn't you just write `[` here?

`{{[[]}}` is a regex `[[]` and that could be written as a non-regex `[`.

What am I missing here that needs the regex? Is `[` special to FileCheck? You have `[options]` there, and that didn't need a regex. Only certain combinations of `[...]` are special?

cs01 replied:
The actual help text looks like this:

```
> ./bin/lldb-server p -h
OVERVIEW: lldb-server platform

USAGE: lldb-server p [options] --listen <[host]:port> [[--] program args...]
```

The regex is only needed with the double brackets, so you were right, there were a couple that were unnecessary. Should be fixed now. 

It's only needed at the `[[--]` part because `[[` is interpreted by FileCheck as variable substitution. For example in `cross-project-tests/dtlto/link-archive-thin.test` PID is substituted:
```
## JSON jobs description.
OUTPUTS-DAG: my.[[PID:[a-zA-Z0-9_]+]].dist-file.json

## Individual summary index files.
OUTPUTS-DAG: start.1.[[PID]].native.o.thinlto.bc{{$}}
OUTPUTS-DAG:   dog.2.[[PID]].native.o.thinlto.bc{{$}}
OUTPUTS-DAG:   foo.3.[[PID]].native.o.thinlto.bc{{$}}
OUTPUTS-DAG:   bar.4.[[PID]].native.o.thinlto.bc{{$}}
```

cs01 commented:
> You could use the CI here to check it. Try enabling the platform mode tests and see what happens.

I removed the `# UNSUPPORTED: system-windows` line at the top and pushed. Is that all I have to do, or do I have to somehow enable platform mode tests manually? And aren't there more tests that run after the PR is merged? Is it possible to run those before landing too?

Otherwise it's ready for another review.

DavidSpickett commented:
> I removed the # UNSUPPORTED: system-windows line at the top and pushed. Is that all I have to do, or do I have to somehow enable platform mode tests manually?

Nothing else to do.

> And aren't there more tests that run after the PR is merged? Is it possible to run those before landing too?

Mostly the same tests on a bunch of platforms but yes there is more that is done. It is technically possible to run them before landing but in this case it's easier to just land it and see what happens. Reverts and fixups are a normal part of "live at main" development.

DavidSpickett commented on lldb/test/Shell/lldb-server/TestPlatformHelp.test:
```diff
@@ -0,0 +1,39 @@
+CHECK: CONNECTION OPTIONS:
+CHECK: --gdbserver-port <port>
+CHECK-SAME: Short form: -P
+CHECK: --listen <{{[[]}}host]:port>
```
Another one here to simplify.

Still wonder why raw `[` doesn't work though.

DavidSpickett reviewed (approved):
Jonas already approved and this LGTM. Whatever makes Filecheck happy is fine with me I was just a bit too curious about the details :)

Do you need someone to merge this for you?

cs01 commented:
> Do you need someone to merge this for you?

Yes I do. Thank you for the review!

cs01 commented:
buildbot failures should be fixed in https://github.com/llvm/llvm-project/pull/164843 fyi @DavidSpickett

adrian-prantl commented:
@cs01 This doesn't seem to work on macOS:

```
[2025-10-23T17:00:02.993Z] ******************** TEST 'lldb-shell :: lldb-server/TestPlatformSuccessfulStartup.test' FAILED ********************
[2025-10-23T17:00:02.993Z] Exit Code: 127
[2025-10-23T17:00:02.993Z] 
[2025-10-23T17:00:02.993Z] Command Output (stdout):
[2025-10-23T17:00:02.993Z] --
[2025-10-23T17:00:02.993Z] # RUN: at line 4
[2025-10-23T17:00:02.993Z] rm -f /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake/lldb-build/tools/lldb/test/Shell/lldb-server/Output/TestPlatformSuccessfulStartup.test.tmp.socket1
[2025-10-23T17:00:02.993Z] # executed command: rm -f /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake/lldb-build/tools/lldb/test/Shell/lldb-server/Output/TestPlatformSuccessfulStartup.test.tmp.socket1
[2025-10-23T17:00:02.993Z] # note: command had no output on stdout or stderr
[2025-10-23T17:00:02.993Z] # RUN: at line 5
[2025-10-23T17:00:02.993Z] timeout 0.2s /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake/lldb-build/bin/lldb-server platform --listen tcp://127.0.0.1:0 --socket-file /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake/lldb-build/tools/lldb/test/Shell/lldb-server/Output/TestPlatformSuccessfulStartup.test.tmp.socket1 > /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake/lldb-build/tools/lldb/test/Shell/lldb-server/Output/TestPlatformSuccessfulStartup.test.tmp.out1 2>&1 || true
[2025-10-23T17:00:02.993Z] # executed command: timeout 0.2s /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake/lldb-build/bin/lldb-server platform --listen tcp://127.0.0.1:0 --socket-file /Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake/lldb-build/tools/lldb/test/Shell/lldb-server/Output/TestPlatformSuccessfulStartup.test.tmp.socket1
[2025-10-23T17:00:02.993Z] # .---command stderr------------
[2025-10-23T17:00:02.993Z] # | 'timeout': command not found
[2025-10-23T17:00:02.993Z] # `-----------------------------
[2025-10-23T17:00:02.993Z] # error: command failed with exit status: 127
[2025-10-23T17:00:02.993Z] 
[2025-10-23T17:00:02.993Z] --
```
https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/lldb-cmake/16164/changes#faf7af864f9258768133894764f1fae58d43bb09

adrian-prantl commented:
Based on the conversation you don't seem to have commit access so I'll go ahead and revert this for you.

cs01 commented:
Would it be possible for me to get commit access?

cs01 commented:
> Why does this need any timeout on top of the LIT timeout?

It doesn't. It's my first LIT and I was trying to be cautious, and not cause any slowdowns, so I added a timeout (which I now understand is not needed). 

I removed it in  https://github.com/llvm/llvm-project/pull/164904 as you suggested. Hopefully that helps it pass on mac.

adrian-prantl commented:
> Thanks for the heads up. I attempted to fix the windows issue in #164843. I am not sure of the normal workflow in cases like this, but in the future it would be easier for me (and probably reviewers) if the change was not reverted entirely and I could try and patch a fix, similar to #164843.

I have no problem with you fixing tests "in production" (within reason) since nobody has access to all the platforms tested by LLVM bots.  But I'd much prefer if you revert the failing commit and then reland it with a fix, rather than layering the fix on top of the previous commit. This also makes it easier to cherry-pick the commit to release branches later on.

Today getting the bots clean was particularly urgent since we had an outage on the arm64 machines and were running on the much slower x86 nodes, so I went ahead and did that.

> Would it be possible for me to get commit access?

https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access

>> Why does this need any timeout on top of the LIT timeout?

> It starts lldb-server and tests that it can successfully listen on a socket, but it must be killed, since it just runs indefinitely. I just removed the test for now to increase the chances of merging without further issue.

Synchronization via timeouts is not really feasible in a CI setting. In order to work with machines under load (or worse, ASANIfied builds of LLDB) the timeout would need to be north of 60s, but then that also means that each successful test would take that long.

In this case it would probably be a better to write this as an API test. You'd want the inner process to, e.g., write to a pipe, wait on that and then deterministically kill the process.

JDevlieghere commented:
It's also worth noting that because of our post-commit testing, heavier use of reverts is a natural part of our development workflow. There's a section in the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy) that sheds some more light on that as it tends to be different from other OSS projects.
